cryptographic_estimators/MAYOEstimator/MAYOAlgorithms/claw_finding.py

```diff
@@ -0,0 +1,142 @@
+
+        EXAMPLES::
+
+            sage: from cryptographic_estimators.MAYOEstimator.mayo_problem import MAYOProblem
+            sage: from cryptographic_estimators.MAYOEstimator.MAYOAlgorithms.claw_finding import ClawFinding
+            sage: E = ClawFinding(MAYOProblem(n=80, m=60, o=18, k=12, q=16))
+            sage: E.X()
+            122.962
+
+        """
+        _, m, _, _, q = self.problem.get_parameters()
+        X = (1 / 2) * (17 - log2(36 * m) + m * log2(q))
+        X_rounded = round(X, 3)
+        return X_rounded
+
+    @optimal_parameter
+    def Y(self):
+        """
+        Return logarithm of the optimal `Y`, i.e. logarithm of no. of hashes to compute
+
+        EXAMPLES::
+
+            sage: from cryptographic_estimators.MAYOEstimator.mayo_problem import MAYOProblem
+            sage: from cryptographic_estimators.MAYOEstimator.MAYOAlgorithms.claw_finding import ClawFinding
+            sage: E = ClawFinding(MAYOProblem(n=80, m=60, o=18, k=12, q=16))
+            sage: E.Y()
+            117.038
+
+        """
+        X =  self.X()
+        _, m, _, _, q = self.problem.get_parameters()
+        Y = max(m * log2(q) - X, 0)
+        Y_rounded = round(Y, 3)
+        return Y_rounded
+
+    def _compute_time_complexity(self, parameters: dict):
+        """
+        Return the time complexity of the algorithm for a given set of parameters
+
+        INPUT:
+
+        - ``parameters`` -- dictionary including the parameters
+
+        TESTS::
+
+            sage: from cryptographic_estimators.MAYOEstimator.mayo_problem import MAYOProblem
+            sage: from cryptographic_estimators.MAYOEstimator.MAYOAlgorithms.claw_finding import ClawFinding
+            sage: E = ClawFinding(MAYOProblem(n=80, m=60, o=18, k=12, q=16))
+            sage: E.time_complexity()
+            134.0384078561605
+
+        """
+        m = self.problem.npolynomials()
+        X = self.X()
+        Y = self.Y()
+        cost_one_hash = self.problem.cost_one_hash
+        time_in_bits = log2(36 * m * 2 ** X + 2 ** Y * 2 ** cost_one_hash)
+        cost_one_field_mult = self.problem.to_bitcomplexity_time(1)
+        time = time_in_bits - cost_one_field_mult
+        return time
+
+    def _compute_memory_complexity(self, parameters: dict):
+        """
+        Return the memory complexity of the algorithm for a given set of parameters
+
+        INPUT:
+
+        - ``parameters`` -- dictionary including the parameters
+
+        TESTS::
+
+            sage: from cryptographic_estimators.MAYOEstimator.mayo_problem import MAYOProblem
+            sage: from cryptographic_estimators.MAYOEstimator.MAYOAlgorithms.claw_finding import ClawFinding
+            sage: E = ClawFinding(MAYOProblem(n=80, m=60, o=18, k=12, q=16))
+            sage: E.memory_complexity()
+            127.038
+
+        """
+        X = self.X()
+        Y = self.Y()
+        n, m, _, k, q = self.problem.get_parameters()
+        mem_evals = log2(log2(q)) + log2(m) + X
```

Comment:
this `+ log(log(q))` is a translation to number of bits stored, even if the complexity_type set to 0

Reply:
You're right, it should be
`mem_evals =  log2(m) + X`  to indicate the log of the number of Fq to be stored for the evaluations of the pk.
`mem_hash = log(256) Y - log(q)` to indicate log of the bit_memory needed to store the hashed interpreted as the number of field elements to store.

Also the hash length (stated as 256) depends on the target security level.
It should be 256, 384 and 512 for categories I, III, and V, respectively.
It can simply use 256 to reduce the complexity of the code? It will underestimate the bit memory by one bit (in the case (`mem_hash` = min(mem_hash, mem_evals)`) 
Or we  include a `security_parameter` for every scheme parameter set? 
In the last case we would use `mem_hash = log(2 * security_parameter) Y - log(q)` .